fix: adapt match function to return true/false instead of the text

EdJoPaTo opened the pull request:
You provided this example in your readme in order to work with localized keyboards:
```js
const { match } = require('telegraf-i18n')

// In case you use custom keyboard with localized labels.
bot.hears(match('keyboard.foo'), (ctx) => ...)
```

This does not work for me. I tried to analyze the reason for it. Your match function returns a function which returns an array with the text in it, it should match in the given language of ctx of the first function. That array would be correct if directly supplied but not with the indirect function. The function should return true or false when matched or not.

So i removed the ternary operator and it worked for me. Please check again why this was made the way it was. Maybe an older version of Telegraf supported this way? Or did I made an error somewhere else?

dotcypress commented:
Hey @EdJoPaTo 
Thank you for your time.

This function need to return array of string, see [Telegraf source](https://github.com/telegraf/telegraf/blob/develop/lib/core/composer.js#L124)

Could you create small bot to reproduce error?

EdJoPaTo commented:
I think I found my issue for the problem I had. Your current code is working for me but I think in a different way then you think it should. (let me explain later in this post)

# My real problem
First, my problem was a interesting combination of telegram clients and my code. I extracted the language from `ctx.from.language_code` on every message and tested my bot with tDesktop and the iOS Telegram Client. My host systems are both english and I changed the settings of the Telegram clients to german in order to test it with a different language.
The Telegram API did now send different `language_code` in some directly following messages because of caching the user language and different client languages I assume.

For example the first message came with an english language code. It created an english keyboard layout. The next message was from a keyboard click. Now the language_code changed. The now german texts did not matched with the english keyboard key click so it did not work.

I think I stopped changing the language of the clients and set it both the same while I started using my wannabe solution. That worked, so I assumed it has to be the problem.
I later changed my source code in order to only load the language from `ctx.from.language_code` from the first `/start` message and then use it cached in the `ctx.session` as there is an example in your readme. Then the language does not flip with every message.

But that is clearly a strange setup with two clients operating at the same time with the bot while changing the language of the clients often.

# The 'problem' I found in your code
I think I understand your thought with the source line you posted there.
But I think you're  missing one step there. (Or maybe I'm wrong, I hope we figure that out)

I seperated it into two parts: First the 'debug' part what I think happens. After that my thoughts about where your thoughts could have went wrong direction.+

## What happens in the code

First, the [.hears Method](https://github.com/telegraf/telegraf/blob/5efc6fb927705cd4e926152193c8da36376cea79/lib/core/composer.js#L109-L111) is used:
```js
bot.hears(match('foo'), ctx => {…})
```
```js
  static hears (triggers, middleware) {
    return Composer.mount('text', Composer.match(normalizeTriggers(triggers), middleware))
  }
```

The trigger given here was `match('foo')` so it will result in this function:
```js
(text, ctx) => (… && text === ctx.i18n.t('foo')) ? [text] : null)
```

The function will then be given to `normalizeTriggers` ([Telegraf source link](https://github.com/telegraf/telegraf/blob/5efc6fb927705cd4e926152193c8da36376cea79/lib/core/composer.js#L204-L223)).
This is the function:
```js
function normalizeTriggers (triggers) {
  if (!Array.isArray(triggers)) {
    triggers = [triggers]
  }
  return triggers.map((trigger) => {
    if (!trigger) {
      throw new Error('Invalid trigger')
    }
    if (typeof trigger === 'function') {
      return trigger
    }
    if (trigger instanceof RegExp) {
      return (value) => {
        trigger.lastIndex = 0
        return trigger.exec(value || '')
      }
    }
    return (value) => trigger === value ? value : null
  })
}
```

First, the function is not an array, so it will be put in one, resulting in this one:


```js
[ (text, ctx) => (… && text === ctx.i18n.t('foo')) ? [text] : null) ]
```
Then, as it is a function, it will stay as it is for now and leave the `normalizeTriggers` function in order to be given to `Composer.match` (as seen above, the function you posted,
 [Telegraf source link](https://github.com/telegraf/telegraf/blob/5efc6fb927705cd4e926152193c8da36376cea79/lib/core/composer.js#L117-L131)):
```js
  static match (triggers, middleware) {
    return Composer.lazy((ctx) => {
      const text = (
        (ctx.message && (ctx.message.caption || ctx.message.text)) ||
        (ctx.callbackQuery && ctx.callbackQuery.data)
      )
      for (let trigger of triggers) {
        ctx.match = trigger(text, ctx)
        if (ctx.match) {
          return middleware
        }
      }
      return Composer.safePassThru()
    })
  }
```
The now function in the array will be given to the for loop.
So `trigger` should be again our function:
```js
(text, ctx) => (… && text === ctx.i18n.t('foo')) ? [text] : null)
```

In the first line of the for loop this function gets executed and the result will be placed in `ctx.match`.
Execution line by line:
```js
// unchanged
(text, ctx) => (… && text === ctx.i18n.t('foo')) ? [text] : null)
// execute i18n.t
(text, ctx) => text === 'hey guys' ? [text] : null
// ternary will result either in
(text, ctx) => [text]
// or
(text, ctx) => null
```
so `ctx.match` will either contain `[text]` or `null`

Then, a simple if follows:
```js
        if (ctx.match) {
          return middleware
        }
```

`ctx.match` with `[text]` does exist and will result in execution of the if content.
`ctx.match` with `null` does not exist, the if content will not be executed.

If the ternary operator will be removed from the source code as this PR is doing either `true` or `false` (or `null` when `!text || !ctx || …`) will be returned.
This will be enough to work.

## The point where I think your thought may have went the wrong way

I think your thought to return an array is there:
The [Telegraf API examples](http://telegraf.js.org/telegraf.html#hears) say it accepts either a string/regex or an array of them (or a function):
```js
telegraf.hears(['foo', 'bar'], (ctx) => {…})
```

when you execute a function on the call like in the following example it still has to return an array:

```js
function foobar() {
  return ['foo', 'bar']
}
telegraf.hears(foobar(), (ctx) => {…})
// results in
telegraf.hears(['foo', 'bar'], (ctx) => {…})
```

But you (indirectly) supply a function which will be executed on time and not on initialization.

```js
function foobar(text) {
  return text === 'foobar'
}
telegraf.hears(foobar, (ctx) => {…})
```

This is what you do, which results in the same as the above one:

```js
function foobar() {
  return function(text) {
    return text === 'foobar'
  }
}
telegraf.hears(foobar(), (ctx) => {…})
```

Note to myself: JavaScript can really produce some not easy to follow code…

---

I hope I explained my thoughts well enough and you can understand what I mean?
If I miss something please point me out.
And sorry for the quite long post. :o

dotcypress commented:
Trigger function must be lazy in case they depends on telegram update, and `match` must be an array just because we need consistency with RegExes, for example:

```js
bot.hears(/insert (\d*)/, (ctx) => {
  console.log(ctx.match[1]) // 
})
```

EdJoPaTo commented:
So it's there in order to keep things the same as with Regex, meant to be used elsewhere too. Thanks for helping me understand it :)

I then thought, it should be separated then. A user (like telegraf-i18n) just using the Telegraf package should not have to case about the continuity of Telegraf and the other way around. As a user I would just expect `true` or `false` to be enough. Maybe `normalizeTriggers` should handle that part and apply `[text]` or `null` based on true or false of the result of the function.

But then things like that would not simply work:
```js
function foobar(text) {
  return /foo(.+)/.exec(text)
}
telegraf.hears(foobar, (ctx) => {…})
```

It has to be an additional check there if the result of the function is true or additional an array... would be to much of a hazzle to check that.

Maybe just add a hint to the documentation that the trigger function takes `(text, ctx)` as input and has to reply a Regex.exec like return value? (Sadly there is no easy to point to Match class in JavaScript). Else I would intuitively just use it wrong.
Maybe introduce the type trigger, that can be `string, string[], Regex, Regex[], function (text, ctx) => Match`. Or add a comment how the function has to look to the `normalizeTriggers` function path and remove it from the documentation as the normal user of Telegraf will likely never use the function way without looking at the source code?

As I think we are on the same page that this PR change is not useful you can close it.

dotcypress commented:
Yes, you are right, documentation is not good enough.
Will update it soon.

And again, thank you for your research 👍
